File: src/test/java/dreamdiary/quiz/api/QuizAddApiTest.java
package dreamdiary.quiz.api;

import dreamdiary.quiz.app.QuizAddRequest;
import dreamdiary.quiz.app.QuizAddUseCase;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Locale;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@ExtendWith(MockitoExtension.class)
class QuizAddApiTest {
    @InjectMocks
    private QuizAddApi quizAddApi;
    @Mock
    private QuizAddUseCase quizAddUseCase;
    private MockMvc mockMvc;

    @BeforeEach
    void setUp() {
        mockMvc = MockMvcBuilders.standaloneSetup(quizAddApi).build();
    }

    @Test
    void addQuiz_status_is_created() throws Exception {
        mockMvc.perform(post("/quizzes")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{}"))
                .andExpect(status().isCreated())
        ;
    }

    @Test
    void addQuiz_passes_data_useCase() throws Exception {
        final String givenTitle = "Quiz Title";
        final String givenContent = "Quiz Content";
        final List<String> givenChoices = List.of("강아지", "고양이", "토끼", "람쥐");
        final LocalDateTime givenReleaseAt = LocalDate.now().plusDays(30L).atTime(14, 0);
        final LocalDateTime givenAnswerReleaseAt = LocalDate.now().plusDays(30L).atTime(16, 30);

        mockMvc.perform(post("/quizzes")
                .contentType(MediaType.APPLICATION_JSON)
                .content("""
                        {
                            "title": "%s",
                            "content": "%s",
                            "releaseAt": "%s",
                            "answerReleaseAt": "%s",
                            "choices": %s
                        }
                        """.formatted(givenTitle, givenContent,
                        givenReleaseAt.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss", Locale.KOREA)),
                        givenAnswerReleaseAt.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss", Locale.KOREA)),
                        givenChoices.stream().map(choice -> "\"" + choice + "\"").toList())
                ));

        final ArgumentCaptor<QuizAddRequest> captor = ArgumentCaptor.forClass(QuizAddRequest.class);

        verify(quizAddUseCase, times(1)).addQuiz(captor.capture());
        assertThat(captor.getValue()).isNotNull();
        assertThat(givenTitle).isEqualTo(captor.getValue().getTitle());
        assertThat(givenContent).isEqualTo(captor.getValue().getContent());
        assertThat(givenChoices).isEqualTo(captor.getValue().getChoices());
        assertThat(givenReleaseAt).isEqualTo(captor.getValue().getReleaseAt());
    }
}

File: presentations/9o-clock-internal-api/src/test/kotlin/dreamdiary/nineoclock/internal/api/quiz/QuizAddApiTest.kt
package dreamdiary.nineoclock.internal.api.quiz

import dreamdiary.nineoclock.application.quiz.usecase.QuizCreateCommand
import dreamdiary.nineoclock.application.quiz.usecase.QuizCreateUseCase
import org.assertj.core.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentCaptor
import org.mockito.Captor
import org.mockito.Mockito
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.reactive.WebFluxTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.http.MediaType
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.test.web.reactive.server.WebTestClient
import org.springframework.web.reactive.function.BodyInserters
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit

private fun <T> capture(argumentCaptor: ArgumentCaptor<T>): T = argumentCaptor.capture()

@ExtendWith(SpringExtension::class)
@WebFluxTest(controllers = [QuizAddApi::class])
internal class QuizAddApiTest {
    @Autowired
    lateinit var webClient:WebTestClient
    @MockBean
    lateinit var mockQuizCreateUseCase: QuizCreateUseCase
    @Captor
    lateinit var commandCaptor: ArgumentCaptor<QuizCreateCommand>

    @Test
    fun addQuiz_status_is_created() {
        val givenRequest = anAddRequest()
        webClient.post()
            .uri("/quizzes")
            .contentType(MediaType.APPLICATION_JSON)
            .body(BodyInserters.fromValue(givenRequest))
            .exchange()
            .expectStatus().isCreated
    }

    @Test
    fun addQuiz_passes_data_to_useCase() {
        val givenRequest = anAddRequest()
        webClient.post()
            .uri("/quizzes")
            .contentType(MediaType.APPLICATION_JSON)
            .body(BodyInserters.fromValue(givenRequest))
            .exchange()

        Mockito.verify(mockQuizCreateUseCase, Mockito.times(1)).createQuiz(capture(commandCaptor))
        assertThat(commandCaptor.value).isNotNull
        assertThat(commandCaptor.value.title).isEqualTo(givenRequest.title)
        assertThat(commandCaptor.value.content).isEqualTo(givenRequest.content)
        assertThat(commandCaptor.value.choices).isEqualTo(givenRequest.choices)
        assertThat(commandCaptor.value.releaseAt.truncatedTo(ChronoUnit.SECONDS)).isEqualTo(givenRequest.releaseAt.truncatedTo(ChronoUnit.SECONDS))
        assertThat(commandCaptor.value.answerReleaseAt.truncatedTo(ChronoUnit.SECONDS)).isEqualTo(givenRequest.answerReleaseAt.truncatedTo(ChronoUnit.SECONDS))
    }

    fun anAddRequest() :QuizAddRequest {
        return QuizAddRequest(
            title = "givenTitle",
            content = "givenContent",
            choices = listOf("A", "B", "C"),
            releaseAt = LocalDateTime.now(),
            answerReleaseAt = LocalDateTime.now()
        )
    }
}
